### PR TITLE
Add support for an externally provided HyperProcess

### LINKED ISSUE #39: Allow user to pass connection / hyper instance
Right now the public facing functions create new Hyper Processes / Connections for the user. This however can be rather expensive (I think 300-400 ms) and limiting for advanced use cases

If someone out there is willing to benchmark and propose an API accordingly would certainly welcome contributions on this

### LINKED ISSUE #51: Allow Telemetry
Right now we explicitly disable Telemetry in the project. If someone cares to I don't have any objection to allow users to share usage stats with Tableau, but at the same time I don't know of the legal implications of doing so

If someone has legal resources at their disposal and has an interest in this would be open to it

### PR DESCRIPTION
**Pull request by vogelsgesang**
So far, pantab always internally created an adhoc `HyperProcess`. This
commit allows users to instead spawn their own `HyperProcess` and pass
it to pantab's functions through a `hyper` keyword argument.

By providing their own `HyperProcess`, users
* have full control over the start-up flags of `HyperProcess` and can,
  e.g., disable the creation of the `hyperd.log` log file.
* reuse the same `HyperProcess` across multiple function calls and
  thereby save a few milliseconds per function invocation.
 
 closes #51, closes #39

### CODE REVIEW

**Comment by vogelsgesang** [comment]
No clue which issue the linux build has with the formatting... On my local machine `black` runs through without any issues.
Could it be that `black` is somehow dependent on the environment it's running on?

**Comment by WillAyd** [comment]
Can you check #118 as well? I'm guessing some of the changes from this PR made their way in there

**Comment by WillAyd** [comment]
> No clue which issue the linux build has with the formatting... On my local machine `black` runs through without any issues.
> Could it be that `black` is somehow dependent on the environment it's running on?

It's possible a versioning thing. I would suggest trying `black-20.8b1` locally to match what CI is getting

**Comment by vogelsgesang** [comment]
Ok - the `black` issues were indeed resolved by installing the right version. Now, I am running into errors from `mypy`:
```
pantab/_tester.py:8: error: Cannot find implementation or library stub for module named 'pytest'
pantab/_hyper_util.py:4: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
pantab/_compat.py:3: error: Cannot find implementation or library stub for module named 'pandas'
pantab/_types.py:4: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
pantab/_writer.py:8: error: Cannot find implementation or library stub for module named 'numpy'
pantab/_writer.py:9: error: Cannot find implementation or library stub for module named 'pandas'
pantab/_writer.py:10: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
pantab/_reader.py:6: error: Cannot find implementation or library stub for module named 'numpy'
pantab/_reader.py:7: error: Cannot find implementation or library stub for module named 'pandas'
pantab/_reader.py:8: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
pantab/tests/test_writer.py:3: error: Cannot find implementation or library stub for module named 'numpy'
pantab/tests/test_writer.py:4: error: Cannot find implementation or library stub for module named 'pandas'
pantab/tests/test_writer.py:5: error: Cannot find implementation or library stub for module named 'pytest'
pantab/tests/test_writer.py:6: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
pantab/tests/test_roundtrip.py:1: error: Cannot find implementation or library stub for module named 'numpy'
pantab/tests/test_roundtrip.py:2: error: Cannot find implementation or library stub for module named 'pandas'
pantab/tests/test_roundtrip.py:3: error: Cannot find implementation or library stub for module named 'pandas.testing'
pantab/tests/test_roundtrip.py:4: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
pantab/tests/test_reader.py:1: error: Cannot find implementation or library stub for module named 'pandas'
pantab/tests/test_reader.py:2: error: Cannot find implementation or library stub for module named 'pandas.testing'
pantab/tests/test_reader.py:3: error: Cannot find implementation or library stub for module named 'pytest'
pantab/tests/test_reader.py:4: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
pantab/tests/conftest.py:3: error: Cannot find implementation or library stub for module named 'numpy'
pantab/tests/conftest.py:3: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
pantab/tests/conftest.py:4: error: Cannot find implementation or library stub for module named 'pandas'
pantab/tests/conftest.py:5: error: Cannot find implementation or library stub for module named 'pytest'
pantab/tests/conftest.py:6: error: Cannot find implementation or library stub for module named 'tableauhyperapi'
```

I don't understand how my code changes could have broken this. Are the CI jobs currently expected to pass?

**Comment by WillAyd** [comment]
They should. The mypy config should ignore missing stubs from external dependencies - not sure why this would appear now. I can take a look in a few hours but should be unrelated to your PR

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by WillAyd** [comment]
Looks like the Mypy issue might just be a bug with the latest version:

https://github.com/python/mypy/issues/9940

I'll push up a patch separately

File: doc/source/examples.rst
Comment (WillAyd) [inline doc/source/examples.rst:116]:
```suggestion
For convenience, pantab's functions internally spawn a `HyperProcess<https://help.tableau.com/current/api/hyper_api/en-us/reference/py/tableauhyperapi.html#tableauhyperapi.HyperProcess>`_. In case you prefer to spawn your own ``HyperProcess``, you can supply it to pantab through the ``hyper_process`` keyword argument.
```

File: pantab/tests/test_roundtrip.py
Comment (WillAyd) [inline pantab/tests/test_roundtrip.py:62]: Can you add to this test a check that a log file is not created? You should be able to use the built in pathlib module to check for that

**Review by WillAyd — COMMENTED** [review]
Looking very nice. mypy issue is fixed on master so feel free to rebase or merge master whenever convenient

File: pantab/tests/test_roundtrip.py
Comment (vogelsgesang) [reply pantab/tests/test_roundtrip.py:66]: done

**Comment by vogelsgesang** [comment]
should be ready to be merged now

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @vogelsgesang

**Comment by farhangithub27** [comment]
Any timeline when this will be released to pypi.

**Comment by vogelsgesang** [comment]
There are still two pull requests in the pipeline (#123 and #128) which we probably want to merge before cutting a new release.

Neither of those changes is a strict requirement before cutting a new release. But we would prefer to have them in the 2.0 release, since at least #123 is also an interface-breaking change. We prefer to bundle those interface-breaking changes into one release instead of having two subsequent interface-breaking changes, so we reduce the burden of upgrading the pantab dependency for down-stream users of pantab.

Originally, we hoped to move #123 and #128 ahead much faster. Unfortunately, it took longer because I first had to get approval by my company to upstream those changes and this whole legal process took a while. But that's out of the way now, so I hope we will get those two changes merged within the next week or so.

The final decision on when to release a new version is up to @WillAyd, though...

**Comment by vogelsgesang** [comment]
@farhangithub27 pantab 2.0 was just released to pypi and this new version also comes with support for the externally provided HyperProcess

**Pull request merged**
